### PR TITLE
Tracking of contexts and contextual commands

### PR DESCRIPTION
**Pull request by gcampax**
Track the current context, pass it to the NLP server.

This is behind a flag, `experimental-contextual-commands`, that must be enabled manually. Otherwise the non-contextual model is used.

### CODE REVIEW

File: test/auto_test_almond.js [hunk 1]
```diff
@@ -3373,9 +3373,7 @@ function main() {
 
     var delegate = new TestDelegate();
 
-    var sempreUrl;
-    if (process.argv[2] !== undefined && process.argv[2].startsWith('--with-sempre='))
-        sempreUrl = process.argv[2].substr('--with-sempre='.length);
+    const sempreUrl = 'https://almond-dev.stanford.edu/nnparser';
```
Comment (sileix) [inline test/auto_test_almond.js:3914]: Maybe it is time to change the name?

File: lib/almond.js [hunk 2]
```diff
@@ -281,6 +286,27 @@ module.exports = class Almond extends events.EventEmitter {
...
+    async setContext(context) {
+        if (context === null) {
+            this._context = {
+                timeout: Infinity,
+                code: 'null',
+                entities: {}
+            };
+        } else {
+            // timeout after 30 seconds
+            let timeout = Date.now() + 30000;
```
Comment (sileix) [inline lib/almond.js]: Shouldn't the time be calculated on the server side?

File: test/auto_test_almond.js (hunk 1 above)
Comment (gcampax) [reply test/auto_test_almond.js:3914]: Probably yeah. Although we don't really have a name for the new parser, do we? It is luinet?

File: lib/almond.js (hunk 2 above)
Comment (gcampax) [reply lib/almond.js]: What server side? The NLP server is completely stateless & anonymous

Comment (sileix) [reply lib/almond.js]: oh, never mind

File: test/auto_test_almond.js (hunk 1 above)
Comment (sileix) [reply test/auto_test_almond.js:3914]: I don't know if we are still gonna change the luinet name. I guess we can wait until we settle down on the name.

**Review by sileix — APPROVED** [review]

**Pull request merged**